test/finitedifftests.jl

```diff
@@ -540,3 +540,11 @@ Base.getindex(x::ImmutableVector, i::Integer) = x.x[i]
         @test J ≈ Matrix(I, 2, 2)
     end
 end
+
+@testset "Hessian Cache test" begin
```

Comment:
A good strategy to catch more bugs like that would be to initialize all caches with randomized versions of `x` in the test suite. That's how I caught it with DifferentiationInterface.